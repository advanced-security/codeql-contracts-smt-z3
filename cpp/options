semmle-extractor-options:--clang -std=c++14 -nostdinc++ -I../../../../include