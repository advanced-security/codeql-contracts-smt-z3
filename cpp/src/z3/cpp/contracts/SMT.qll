import z3.cpp.contracts.SpecificationCase
import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis

module Contracts {
  bindingset[exp, lb, ub]
  private string boundsToAssert(string exp, float lb, float ub) {
    result = upperBoundsToAssert(exp, ub) + "\n" + lowerBoundsToAssert(exp, lb) + "\n"
  }

  bindingset[exp, ub]
  private string upperBoundsToAssert(string exp, float ub) {
    result = "(assert (<= " + exp + " " + ub + ")) ; establishment of upper bounds"
  }

  bindingset[exp, lb]
  private string lowerBoundsToAssert(string exp, float lb) {
    result = "(assert (>= " + exp + " " + lb + ")) ; establishment of lower bounds"
  }

  bindingset[exp, categoricalValue]
  private string categoricalValueToAssert(string exp, string categoricalValue) {
    result = "(assert (= " + exp + " " + categoricalValue + ")) ; establishment categorical value"
  }

  /**
   * This predicate works differently than the way that preconditions are constructed.
   * In the case of preconditions the scope is limited to the precondition's scope.
   * The constants and declarations however are exposed at the script level. For
   * this reason we must de-duplicate the declarations.
   *
   * There are two types of declarations we include. Those that are directly needed
   * in the scope of the contract and those that are needed in the body of the
   * callsite.
   */
  string extractDeclarations(CodeQLSpecificationCase specCase, FunctionCall fc) {
    // handles the declarations inside of the specified method, including
    // the ones stored in the spec database.
    specCase.getASpecificationClause().getSMTDeclarations() = result
    or
    // handles the declarations possibly needed by injecting scoped
    // variables
    fc.getTarget() = specCase.getFunction() and
    // analyze each argument
    exists(
      BoundaryAnalyzedFunctionCall bfc, ContractParameter p, Z3::Sort z3Sort, Expr valueExpr,
      Expr anchorExpr, string id
    |
      fc = bfc and
      valueExpr = bfc.getAValueExpression(anchorExpr, p) and
      // pull the type from the parameter if it's the value
      // being passed to the function
      (
        if valueExpr = anchorExpr
        then (
          z3Sort = z3SortForQlRawType(p.getType()) and id = p.toId()
        ) else (
          z3Sort = z3SortForQlType(valueExpr) and
          id = valueExpr.(TerminalVariableAccess).toId(anchorExpr, p)
        )
      ) and
      result = "(declare-const " + id + " " + z3Sort.toString() + ")"
    )
  }

  string concatDeclarations(CodeQLSpecificationCase specCase, FunctionCall fc) {
    fc.getTarget() = specCase.getFunction() and
    result = concat(string s | s = extractDeclarations(specCase, fc) | s + "\n")
  }

  string assertPreconditionBody(CodeQLSpecificationClause clause) {
    result = concat(CodeQLSpecificationClause c | c = clause | (c.getSMTAsserts() + "\n"))
  }

  string getPrecondition(CodeQLSpecificationCase specCase) {
    exists(int preconditionNumber, string pcBody |
      pcBody =
        rank[preconditionNumber](string pc |
          pc = assertPreconditionBody(specCase.getASpecificationClause())
        |
          pc
        ) and
      result =
        "(define-fun precondition_" + preconditionNumber.toString() + "() Bool \n\t" + pcBody +
          ")\n\n"
    )
  }

  string concatPreconditions(CodeQLSpecificationCase specCase) {
    result = concat(getPrecondition(specCase) + "\n")
  }

  /**
   * Boundary analysis is done on a per argument basis to the function call.
   * We do this by mapping the formal parameter names in SMT to the values
   * produced by range analysis at the call site for its actual parameters.
   */
  string getBounds(CodeQLSpecificationCase specCase, FunctionCall fc) {
    // link these together
    fc.getTarget() = specCase.getFunction() and
    // analyze each argument
    exists(
      BoundaryAnalyzedFunctionCall bfc, ContractParameter p, Z3::Sort z3Sort, Expr valueExpr,
      Expr anchorExpr, string id
    |
      fc = bfc and
      valueExpr = bfc.getAValueExpression(anchorExpr, p) and
      // pull the type from the parameter if it's the value
      // being passed to the function
      (
        if valueExpr = anchorExpr
        then (
          z3Sort = z3SortForQlRawType(p.getType()) and id = p.toId()
        ) else (
          z3Sort = z3SortForQlType(valueExpr) and
          id = valueExpr.(TerminalVariableAccess).toId(anchorExpr, p)
        )
      ) and
      // it isn't necessary to break up these cases --
      // we could just take a disjunction but this is presented
      // this way for clarity that we are taking different approaches
      // to value analysis.
      if z3Sort = Z3::TOptionNullSort()
      then
        exists(string categoricalValue |
          bfc.getACategoricalValue(valueExpr, anchorExpr, p) = categoricalValue and
          result = categoricalValueToAssert(id, categoricalValue)
        )
      else
        exists(float lb, float ub |
          bfc.getALowerBound(valueExpr.getParent*(), anchorExpr, p) = lb and
          ub = bfc.getAUpperBound(valueExpr.getParent*(), anchorExpr, p) and
          result = boundsToAssert(id, lb, ub)
        )
    )
  }

  string concatBounds(CodeQLSpecificationCase specCase, FunctionCall fc) {
    fc.getTarget() = specCase.getFunction() and
    result = concat(string s | s = getBounds(specCase, fc) | s + "\n")
  }

  /**
   * The goal of contract checking is not to find an assignment of variables
   * that allows a formula to be true; the goal is to find an assignment wherein
   * the formula is false. Therefore to make this work with a SAT solver
   * (which finds assignments that make a formula true) we use it in reverse
   * and ask the solver for a cases wherein the solver searches for an
   * assignment to an invalid form of the formula.
   *
   * Stated more concisely, for a formula $F$ over variables $V$ a counter
   * example is an assignment of variables $V$ for which $\neg F$ is satisfiable.
   */
  string getCounterExamples(CodeQLSpecificationCase specCase) {
    exists(string preconditionList, int numberOfPreconditions |
      // note -- we translate to the precondition body here because it is
      // possible to have duplicate preconditions. These are automatically
      // filtered out by set equality so to replicate that behavior here,
      // the conversion via `assertPreconditionBody` ensures that the
      // deduplication takes place.
      numberOfPreconditions = count(assertPreconditionBody(specCase.getASpecificationClause())) and
      preconditionList =
        concat(int idx | idx = [1 .. numberOfPreconditions] | "precondition_" + idx.toString() + " ")
            .trim() and
      if numberOfPreconditions = 1
      then result = "(assert (not " + preconditionList + "))"
      else result = "(assert (not (and " + preconditionList + ")))"
    )
  }

  string toSMT(CodeQLSpecificationCase specCase, FunctionCall fc) {
    // scope this down to a single function
    fc.getTarget() = specCase.getFunction() and
    // do the boundary analysis
    result =
      "                     \n" + ";;                            \n" +
        ";; Declarations               \n" +
        "(declare-datatypes () ((OptionNullSort MaybeNull NotNull)))\n\n" +
        concatDeclarations(specCase, fc) + ";;                            \n" +
        ";; Precondition Definition    \n" + concatPreconditions(specCase) +
        ";;                            \n" + ";; Boundary/Context Assertions\n" +
        concatBounds(specCase, fc) + ";;                            \n" +
        ";; Counter Example Assertion  \n" + getCounterExamples(specCase)
  }
}
