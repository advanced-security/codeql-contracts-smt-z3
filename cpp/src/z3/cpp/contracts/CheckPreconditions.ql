/**
 * @id cpp/contracts/check-preconditions
 * @name Check Preconditions
 * @description Check the preconditions of called methods and report possible
 *              violations.
 * @kind problem
 * @precision high
 * @problem.severity warning
 * @tags cpp/contracts/check-preconditions
 *       codeql/contracts/preconditions/automated
 */

import cpp
import z3.cpp.contracts.SpecificationCase
import z3.cpp.contracts.SMT
import z3.Z3
// enable comment specs
import z3.cpp.contracts.style.AnnotatedCAssertStyleSpecifiedElement

from FunctionCall fc, CodeQLSpecificationCase specCase, string proof, string ce
where
  specCase.getFunction() = fc.getTarget() and
  proof = Contracts::toSMT(specCase, fc) and
  ce = Z3::getModel(proof)
select fc,
  "The $@ at guarding this function call may not be satisfied by this implementation. The SMT Solver found the following model which violates this method's precondition: \n\n"
    + ce, specCase, "specification case"
