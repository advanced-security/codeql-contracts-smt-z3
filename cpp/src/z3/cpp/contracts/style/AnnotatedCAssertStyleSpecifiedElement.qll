import cpp
import z3.cpp.contracts.SpecificationCase

/**
 * Describes a C-style assert that has been annotated with a comment indicating
 * it should be treated as a contract specification.
 *
 * For example:
 *
 * ```
 *   /*@ requires @* /
 *   assert(x > 0);
 * ```
 */
class AnnotatedCAssertStyleSpecifiedElement extends SpecifiedElement, MacroInvocation {
  Expr body;
  string keyword;
  string contents;
  Comment exprComment;

  AnnotatedCAssertStyleSpecifiedElement() {
    // model the LHS of a logical or style CAssert
    exists(LogicalOrExpr b |
      body = b.getLeftOperand() and
      b = this.getAGeneratedElement() and
      not exists(LogicalOrExpr e1 |
        not e1 = b and
        b.getParent*() = e1
      )
    ) and
    exprComment.getCommentedElement() = this.getExpr().getEnclosingStmt() and
    (
      // Find comments such as '/*@ requires @*/' or '/*@ ensures @*/'.
      contents =
        exprComment
            .getContents()
            .splitAt("\n")
            .regexpFind("^\\s*\\/\\*@\\s+requires\\s+@\\*\\/$", _, _) and
      keyword = "requires"
      or
      contents =
        exprComment
            .getContents()
            .splitAt("\n")
            .regexpFind("^\\s*\\/\\*@\\s+ensures\\s+@\\*\\/$", _, _) and
      keyword = "ensures"
      or
      contents =
        exprComment
            .getContents()
            .splitAt("\n")
            .regexpFind("^\\s*\\/\\*@\\s+assume\\s+@\\*\\/$", _, _) and
      keyword = "assume"
    )
  }

  override string getSource() { getBody().toString() = result }

  override Expr getBody() { result = body }

  override Function getFunction() { result = getBody().getEnclosingFunction() }

  override predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    // for a comment based assert specification the location is essentially
    // the start of the comment to the end of the assert.
    exprComment.getLocation().hasLocationInfo(filepath, startline, startcolumn, _, _) and
    this.(LibcAssert).getLocation().hasLocationInfo(_, _, _, endline, endcolumn)
  }

  string getContents() { result = contents }

  override string getKeyword() { result = keyword }
}
