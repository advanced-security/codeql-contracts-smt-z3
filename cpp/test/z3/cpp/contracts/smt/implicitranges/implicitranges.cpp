#include "assert.h"
#include <cstddef>

class A {
public:
    int *a;
    int d;
    A *b;
    int operator==(A x)
    {
        return 0;
    }
};

void f1(int i, A* a, A &b){

    /*@ requires @*/
    assert(a);

    /*@ requires @*/
    assert(a->a);

    /*@ requires @*/
    assert(a->d > 0);

    /*@ requires @*/
    assert(b.a);
        
    /*@ requires @*/
    assert(a->b->a);

    /*@ requires @*/
    assert(b.b->a);

    /*@ requires @*/
    assert(b.b->b->b->b->b);

    /*@ requires @*/
    assert(a->b->b->b->b->b);
}

void f2(){
    int i;
    A* c;
    A  d;

    *c->a = 100; // DNE in contract (contract is reference)
    *c->a = 109; // DNE in contract (contract is reference)
    
    c->d = 101;  // Exists in contract 
    c->d = 1011; // Exists in contract 

    *c->b->a = 102; // DNE in contract (contract is a reference)
    c->b->b->b->b->b = NULL; // Exists in contract (same type)
    
    *d.a = 103; // DNE in contract (contract is a reference)
    *d.b->a = 104; // DNE in contract (no reference)
    d.b->b->b->b->b = NULL; // Exists in contract (is a reference)

    i = 10; // DNE in contract 

    f1(i, c, d);
}