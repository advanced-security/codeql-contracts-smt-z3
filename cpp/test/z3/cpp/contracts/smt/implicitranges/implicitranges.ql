/**
 * @id cpp/test/SMT/Identifiers
 * @name Identifiers
 * @description Checks how identifiers are encoded in SMT.
 */

import cpp
import z3.cpp.contracts.SpecificationCase
import z3.cpp.contracts.SMT
import z3.cpp.contracts.style.AnnotatedCAssertStyleSpecifiedElement
import z3.Z3

string getBounds2(CodeQLSpecificationCase specCase, FunctionCall fc, Z3::Sort z3Sort) {
  // link these together
  fc.getTarget() = specCase.getFunction() and
  // analyze each argument
  exists(
    BoundaryAnalyzedFunctionCall bfc, ContractParameter p, Expr valueExpr, Expr anchorExpr,
    string id
  |
    fc = bfc and
    valueExpr = bfc.getAValueExpression(anchorExpr, p) and
    // pull the type from the parameter if it's the value
    // being passed to the function
    (
      if valueExpr = anchorExpr
      then (
        z3Sort = z3SortForQlRawType(p.getType()) and id = p.toId()
      ) else (
        z3Sort = z3SortForQlType(valueExpr) and
        id = valueExpr.(TerminalVariableAccess).toId(anchorExpr, p)
      )
    ) and
    //result = id
    // it isn't necessary to break up these cases --
    // we could just take a disjunction but this is presented
    // this way for clarity that we are taking different approaches
    // to value analysis.
    if z3Sort = Z3::TOptionNullSort()
    then
      exists(string categoricalValue |
        bfc.getACategoricalValue(valueExpr, anchorExpr, p) = categoricalValue and
        //and result = categoricalValueToAssert(id, categoricalValue)
        result = id + "CATEGORY INFER"
      )
    else
      exists(float lb, float ub |
        bfc.getALowerBound(valueExpr.getParent*(), anchorExpr, p) = lb and
        ub = bfc.getAUpperBound(valueExpr.getParent*(), anchorExpr, p) and
        //result = boundsToAssert(id, lb, ub)
        // lb = 0 and //ub = 0 and
        // ub = upperBound(valueExpr.getParent*().(Expr).getFullyConverted()) and
        result = id + "= NUMERIC INFER"
      )
  )
}

string testSubstituteRoot(
  CodeQLSpecificationCase specCase, FunctionCall fc, string orig, string anchorId
) {
  // link these together
  fc.getTarget() = specCase.getFunction() and
  // analyze each argument
  exists(
    BoundaryAnalyzedFunctionCall bfc, ContractParameter p, Z3::Sort z3Sort, Expr valueExpr,
    Expr anchorExpr, string id
  |
    fc = bfc and
    valueExpr = bfc.getAValueExpression(anchorExpr, p) and
    // pull the type from the parameter if it's the value
    // being passed to the function
    (
      if valueExpr = anchorExpr
      then (
        z3Sort = z3SortForQlRawType(p.getType()) and id = p.toId()
      ) else (
        z3Sort = z3SortForQlType(valueExpr) and
        id = valueExpr.(TerminalVariableAccess).toId(anchorExpr, p)
      )
    ) and
    id = result and
    orig = valueExpr.(TerminalVariableAccess).toId() and
    anchorId = anchorExpr.(TerminalVariableAccess).toId()
  )
}

predicate testValueExpressions(
  FunctionCall fc, CodeQLSingleSpecificationCase specCase, Expr anchorExpr, Expr expr
) {
  fc.getTarget() = specCase.getFunction() and
  exists(BoundaryAnalyzedFunctionCall bfc, ContractParameter p |
    fc = bfc and
    expr = bfc.getAValueExpression(anchorExpr, p)
  )
}

TerminalVariableAccess testGetAValueExpression(
  BoundaryAnalyzedFunctionCall bfc, Expr fcExpr, ContractParameter cp
) {
  bfc.getAnActualFormalParameterPair(fcExpr, cp) and
  (
    // an implicit reference that is the most recent one
    result = getAPrefixedVariableReferenceForExpr2(fcExpr) and
    isLastVariableReferenceInContext(result)
    or
    // an explicit one
    result = fcExpr
  )
}

TerminalVariableAccess getAPrefixedVariableReferenceForExpr2(Expr exp) {
  // first, convert the target into the normalized form. We
  // do this but if no conversion exists (because there isn't one
  // for example, in the case of an arithmetic expression) then the
  // `variableAccessesForExpression` predicate is empty.
  exists(TerminalVariableAccess parentVa |
    parentVa.forExpr() = exp and
    exists(TerminalVariableAccess childVa |
      // see if the parent is a prefix of the child.
      parentVa.refersTo(childVa) and
      // but not identical
      not childVa.toId() = parentVa.toId() and
      result = childVa
    )
  )
}

predicate testRefersTo(
  TerminalVariableAccess va1, TerminalVariableAccess va2, string id1, string id2
) {
  id1 = va1.toId() and
  id2 = va2.toId() and
  va1.refersTo(va2)
}

predicate testNotRefersTo(
  TerminalVariableAccess va1, TerminalVariableAccess va2, string id1, string id2
) {
  id1 = va1.toId() and
  id2 = va2.toId() and
  not va1.refersTo(va2)
}

predicate testStripPrefix(TerminalVariableAccess va1, string prefix, string stripped, string id) {
  va1.stripPrefix(prefix) = stripped and id = va1.toId()
}

predicate testStripPrefix2(
  TerminalVariableAccess va1, TerminalVariableAccess va2, string id1, string id2, string stripped1,
  string stripped2
) {
  va1.stripPrefix(_) = stripped1 and
  id1 = va1.toId() and
  va2.stripPrefix(_) = stripped2 and
  id2 = va2.toId()
}

predicate testTerminalVariableAccessId(TerminalVariableAccess va, string id) { id = va.toId() }

predicate testTerminalVariableAccessIdExpr(TerminalVariableAccess va, Expr exp, string id) {
  va.forExpr() = exp and id = va.toId()
}

Z3::Sort testTerminalVariableAccessTypes(TerminalVariableAccess va) {
  z3SortForQlRawType(va.getTarget().getType()) = result
}

Type testTerminalVariableAccessType(TerminalVariableAccess va, string clazz) {
  result = va.getTarget().getType() and
  result.getAQlClass() = clazz
}

predicate testBoundaryAnalyzedFunctionCall(BoundaryAnalyzedFunctionCall bfc) { any() }

predicate testContractParameter(ContractParameter cp) { any() }

predicate testContractPairs(BoundaryAnalyzedFunctionCall bfc, Expr expr, ContractParameter p) {
  bfc.getAnActualFormalParameterPair(expr, p)
}

Z3::Sort z3SortForQlType2(Type t, VariableAccess va) {
  (
    t instanceof PointerType and dereferenced(va) and supportedPrimitiveType(t.stripType(), result)
    or
    (t instanceof PointerType and not dereferenced(va)) and
    result instanceof Z3::OptionNullSort
    or
    supportedPrimitiveType(t, result)
  )
}

Z3::Sort z3SortForQlType3(Type t, TerminalVariableAccess va) {
  va.getTarget().getType() = t and
  (
    t instanceof PointerType and dereferenced(va) and supportedPrimitiveType(t.stripType(), result)
    or
    (t instanceof PointerType and not dereferenced(va)) and
    result instanceof Z3::OptionNullSort
    or
    supportedPrimitiveType(t, result)
  )
}

/// note: predicates above are for debugging
from FunctionCall fc, CodeQLSpecificationCase specCase, string bounds
where
  fc.getTarget() = specCase.getFunction() and
  Contracts::getBounds(specCase, fc) = bounds
// getBounds2(specCase, fc) = bounds
select fc, specCase, bounds

// check that the preconditions are getting generated
query string checkPreconditions(CodeQLSpecificationCase specCase) {
  Contracts::concatPreconditions(specCase) = result
}

// check that the declarations are getting generated
query string checkDeclarations(FunctionCall fc, CodeQLSpecificationCase specCase) {
  fc.getTarget() = specCase.getFunction() and
  Contracts::extractDeclarations(specCase, fc) = result
}
