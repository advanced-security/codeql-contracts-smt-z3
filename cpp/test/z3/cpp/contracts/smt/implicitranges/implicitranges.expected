WARNING: unused predicate 'getBounds2' (implicitranges.ql:13,8-18)
WARNING: unused predicate 'testBoundaryAnalyzedFunctionCall' (implicitranges.ql:173,11-43)
WARNING: unused predicate 'testContractPairs' (implicitranges.ql:177,11-28)
WARNING: unused predicate 'testContractParameter' (implicitranges.ql:175,11-32)
WARNING: unused predicate 'testGetAValueExpression' (implicitranges.ql:97,24-47)
WARNING: unused predicate 'testNotRefersTo' (implicitranges.ql:136,11-26)
WARNING: unused predicate 'testRefersTo' (implicitranges.ql:128,11-23)
WARNING: unused predicate 'testStripPrefix' (implicitranges.ql:144,11-26)
WARNING: unused predicate 'testStripPrefix2' (implicitranges.ql:148,11-27)
WARNING: unused predicate 'testSubstituteRoot' (implicitranges.ql:58,8-26)
WARNING: unused predicate 'testTerminalVariableAccessId' (implicitranges.ql:158,11-39)
WARNING: unused predicate 'testTerminalVariableAccessIdExpr' (implicitranges.ql:160,11-43)
WARNING: unused predicate 'testTerminalVariableAccessType' (implicitranges.ql:168,6-36)
WARNING: unused predicate 'testTerminalVariableAccessTypes' (implicitranges.ql:164,10-41)
WARNING: unused predicate 'testValueExpressions' (implicitranges.ql:87,11-31)
WARNING: unused predicate 'z3SortForQlType2' (implicitranges.ql:181,10-26)
WARNING: unused predicate 'z3SortForQlType3' (implicitranges.ql:192,10-26)
#select
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_a 109)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_a 109)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_a 109)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_a -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_a 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_a 109)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_a 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_a -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_b_qlcontracts_a 102)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_b_qlcontracts_a 102)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_b_qlcontracts_a 102)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_b_qlcontracts_a -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_b_qlcontracts_a 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_b_qlcontracts_a 102)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_b_qlcontracts_a 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_b_qlcontracts_a -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_d 1011)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_d 1011)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_d 1011)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_d -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_d 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_d 1011)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__a_qlcontracts_d 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__a_qlcontracts_d -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__b_qlcontracts_a 103)) ; establishment of upper bounds\n(assert (>= _valueType__b_qlcontracts_a 103)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__b_qlcontracts_a 103)) ; establishment of upper bounds\n(assert (>= _valueType__b_qlcontracts_a -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__b_qlcontracts_a 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__b_qlcontracts_a 103)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__b_qlcontracts_a 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__b_qlcontracts_a -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__b_qlcontracts_b_qlcontracts_a 104)) ; establishment of upper bounds\n(assert (>= _valueType__b_qlcontracts_b_qlcontracts_a 104)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__b_qlcontracts_b_qlcontracts_a 104)) ; establishment of upper bounds\n(assert (>= _valueType__b_qlcontracts_b_qlcontracts_a -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__b_qlcontracts_b_qlcontracts_a 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__b_qlcontracts_b_qlcontracts_a 104)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__b_qlcontracts_b_qlcontracts_a 2147483647)) ; establishment of upper bounds\n(assert (>= _valueType__b_qlcontracts_b_qlcontracts_a -2147483648)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (<= _valueType__i 10)) ; establishment of upper bounds\n(assert (>= _valueType__i 10)) ; establishment of lower bounds\n |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (= _referenceType__a MaybeNull)) ; establishment categorical value |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (= _referenceType__a_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b MaybeNull)) ; establishment categorical value |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (assert (= _referenceType__b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b MaybeNull)) ; establishment categorical value |
checkPreconditions
| implicitranges.cpp:15:6:15:7 | specification of f1 | (define-fun precondition_1() Bool \n\t(>  _valueType__a_qlcontracts_d  0 ) \n)\n\n\n(define-fun precondition_2() Bool \n\t(not (=  _referenceType__a MaybeNull))\n)\n\n\n(define-fun precondition_3() Bool \n\t(not (=  _referenceType__a_qlcontracts_a MaybeNull))\n)\n\n\n(define-fun precondition_4() Bool \n\t(not (=  _referenceType__a_qlcontracts_b_qlcontracts_a MaybeNull))\n)\n\n\n(define-fun precondition_5() Bool \n\t(not (=  _referenceType__a_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b MaybeNull))\n)\n\n\n(define-fun precondition_6() Bool \n\t(not (=  _referenceType__b_qlcontracts_a MaybeNull))\n)\n\n\n(define-fun precondition_7() Bool \n\t(not (=  _referenceType__b_qlcontracts_b_qlcontracts_a MaybeNull))\n)\n\n\n(define-fun precondition_8() Bool \n\t(not (=  _referenceType__b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b MaybeNull))\n)\n\n\n |
checkDeclarations
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _referenceType__a OptionNullSort) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _referenceType__a_qlcontracts_a OptionNullSort) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _referenceType__a_qlcontracts_b_qlcontracts_a OptionNullSort) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _referenceType__a_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b OptionNullSort) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _referenceType__b_qlcontracts_a OptionNullSort) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _referenceType__b_qlcontracts_b_qlcontracts_a OptionNullSort) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _referenceType__b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b_qlcontracts_b OptionNullSort) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _valueType__a_qlcontracts_a Int) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _valueType__a_qlcontracts_b_qlcontracts_a Int) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _valueType__a_qlcontracts_d Int) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _valueType__b_qlcontracts_a Int) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _valueType__b_qlcontracts_b_qlcontracts_a Int) |
| implicitranges.cpp:62:5:62:6 | call to f1 | implicitranges.cpp:15:6:15:7 | specification of f1 | (declare-const _valueType__i Int) |
